### DISCUSSION TITLE
A bunch of new long-term ideas for the project

### DISCUSSION THREAD

**Discussion by amtoine** [phenax/bsp-layout#48] [Ideas]
As i was going through the issues and the PRs to get a gist of the current state of the project, i had the following ideas.
These are mainly enhancement propositions, i've already directly opened issues for the bugs i encountered! :+1: 

**DISCLAIMER**: these ideas are not of a high priority but would, to me at least, improve the experience a bit, from the point of view of either a user or a dev. :blush: 

**THE IDEAS**: 
  - **`bsp-layout get` all the desktops**: add the possibility to pass an `all` argument to `bsp-layout get` to get a list of the layouts of each and everyone of the desktops in the current `bspwm` session. That would essentially be an alias to something like
```bash
for desktop in $(bspc query --desktops --names); do
  layout=$(bsp-layout get $desktop)
  echo "desktop $desktop: $layout"
done
```
  - **explicitely operate on the current desktop**: simply add the possibility to explicitely give `current` as an argument to the subcommands of `bsp-layout` in addition to the implicite way with no extra argument, e.g. make `bsp-layout get` and `bsp-layout get current` equivalent
  - **rename the command**: in the realm of the "*Binary Space Partitioning Window Manager*", we have (1) `bspwm`, the socket of the manager itself, and (2) `bspc`, the controller of the socket. I think it would be more coherent with the base project to ship `bsp-layout` and `bsp-layout-git` with the `bspl` command instead of the current `bsp-layout` command, which essentially means to modify the `Makefile`! For backcompatibility concerns, the packages should also come with the `bsp-layout` command which would only call `bspl` after printing a deprecation warning to `stdout`.
  - **shell completion**: no idea how to implement that right now, but i think it would be sooo cool to be able to autocomplete:
    + (1) the subcommands of `bsp-layout`, e.g. `get` and `remove`
    + (2) even the name of the layouts after `once` or `set`, e.g. `monocle` and `even`
    + (3) the names of the desktops, e.g. `1`, `2`, ... after `get`
  - **use the full switches of `bspc`**: to help new users and new devs to get up to speed with the source code and especially complex commands like `bspc`, that could be a usefull thing to use the full switches of the `bspc` command, e.g.  replace `bspc query -D` by `bspc query --desktops`
  - **restore the previous layout**: when registering a new layout with `bsp-layout` , the idea would be to save the previous native layout used by `bspwm`, e.g.  `alternate`, `spiral` or `longest_side` according to [the README of `bspwm`](https://github.com/baskerville/bspwm#automatic-mode). When the user would unregister, i.e. `bsp-layout remove`, the layout, the previous one would be restored and the windows rearranged accordingly.
  - **run unit tests when opening a PR / before merging branches**: not sure if that is possible as `bspwm` requires a graphical server to run, but having a workflow in `~/.github` would help catch some bugs. even just a spellcheck!
  - **add issue templates**: to help the users and the devs open new bug reports and feature requests, *GitHub* templates can be very helpful!
  - **add even more doc**: there is definitely documentation in the code base of `bsp-layout`! however, i think there is never really enough documentation :wink: more comments, docstrings for functions, to explain the arguments, etc, etc... that kind of stuff. no real plan on that :slightly_smiling_face: 
  - **separate building and installation**: i do like the idea that `bsp-layout` "builds" the `bash` source code into a usable script. the use of `{{VERSION}}` and `{{SOURCE_PATH}}` for instance is very smart to change the behaviour of the final program though the `Makefile`! However, from a dev point of view, or even someone just trying to build the application from source, that is quite overkill to be only able to `sudo make install` :thinking: you need to enter the user password and possibly mess around with the system installation of the script... that would be nice to have some more layers of installation. e.g. `make` or `make build` or `make bsp-layout` to "build" the script, namely giving access to a usable executable `bsp-layout` in the root of the repo, `make clean` to remove "build" files and `sudo make install` to finally install the built script.

Hope you'll find these ideas interesting, usefull in some way!
Again, these are not top priority as they're only new extra features, just things i thought of for the long run of the project! :wink:

cheers @phenax :yum: :wave:

**Comment by phenax** [comment]
These are some really good suggestions! Some of the points like `explicitely operate on the current desktop`, `bsp-layout get all the desktops`, `rename the command`, `use the full switches of bspc`, `add issue templates`, `add even more doc` seem pretty trivial so we can pick them up as we go.


> rename the command

I think we should maintain both `bspl` and `bsp-layout` together without deprecation because a lot of people are using it inside their config and keybindings so those deprecation warnings will never be noticed.

> run unit tests

Good idea! Not sure how useful bash unit tests will be but worth a shot. At the very least, it will give us a tiny safety net against introducing new bugs.

> separate building and installation

We can already create a build output with this -> `make PREFIX=$PWD/output install` and run it as `./output/bin/bsp-layout help`

> shell completion

Not sure about this one since it adds very little value and increases the surface area of changes we need to maintain. Doesn't seem worth the effort.

Feel free to create low priority issues for these (except for shell completion which I'm not convinced is worth it).

**Reply by amtoine** [reply]
> These are some really good suggestions!
>
glad you like them :blush: 

> Some of the points like `explicitely operate on the current desktop`, `bsp-layout get all the desktops`, `rename the command`, `use the full switches of bspc`, `add issue templates`, `add even more doc` seem pretty trivial so we can pick them up as we go.
> 
yep definitely!
btw, adding a project could be cool to track issues, PRs, reviews, things to do, and so on :+1: 

> I think we should maintain both `bspl` and `bsp-layout` together without deprecation because a lot of people are using it inside their config and keybindings so those deprecation warnings will never be noticed.
> 
yes, we should keep the `bsp-layout` command as well!
but it could only be as simple as
```bash
# bsp-layout
bash /usr/local/lib/bsp-layout/layout.sh
```
whereas `bspl` could be the "real" script :thinking: 

> Good idea! Not sure how useful bash unit tests will be but worth a shot. At the very least, it will give us a tiny safety net against introducing new bugs.
> 
exactly the purpose of such tests! :wink: 

> We can already create a build output with this -> `make PREFIX=$PWD/output install` and run it as `./output/bin/bsp-layout help`
> 
oooh really?
that's cool :+1: 
no particular interest in `make`, `make bsp-layout` and `make clean` targets? :thinking: 
i was more thinking about a "real" separation, to install in two steps. here, the `sudo make install` that follows `make ...` will build the script again...
and that would, to me again, be a lot easier and make more sense to build the script locally, in the root of the repo, with either `make` or `make {name_of_program}`, at least from a dev point ot view or just someone wanting to test the script in the blink of an eye :yum: 

> Not sure about this one since it adds very little value and increases the surface area of changes we need to maintain. Doesn't seem worth the effort.
> 
ok, that is true :confused: 
maybe i'll reserve that for a third party repo, like a plugin to use with `bsp-layout` :+1: 

> Feel free to create low priority issues for these (except for shell completion which I'm not convinced is worth it).
>
i'll create issues as we go :yum: 

not sure i can add labels to issues though :confused:
and i would have self-assigned such issues as well :thinking:

**Comment by amtoine** [comment]
to keep track of the ideas and their progress :yum: 

**legend**:
- :x: won't work on that
- :white_check_mark: approved
- :arrow_right: work in progress
- :hand: review in progress
- :heavy_check_mark:  done
- :question: not fully approved nor rejected

**the ideas**:
- :question: *restore the previous layout*
- :question: *separate building and installation*
- :white_check_mark: *add debug statement and `make` options*
- :white_check_mark: *refactor the install process*
- :white_check_mark: *`bsp-layout get` all the desktops*
- :white_check_mark: *explicitely operate on the current desktop*
- :arrow_right: *rename the command*
- :arrow_right: *use the full switches of `bspc`*
- :white_check_mark: *run unit tests when opening a PR / before merging branches*
- :heavy_check_mark: *add issue templates*
- :arrow_right: *add even more doc*
- :x: *shell completion*: probably will move to [a2n-s/bsp-layout-completion](https://github.com/a2n-s/bsp-layout-completion) some day

**Comment by amtoine** [comment]
i thought about something, a new idea

**generate different versions**: right now, `bsp-layout version` only gives the version in the `Makefile`, e.g. `0.0.10-1`. I see two potential problems with that:
- this has to be harcoded, so that is pretty error prone...
- and it shows no difference at all between the installation (1) of the more stable `bsp-layout` from the AUR, (2) the less stable `bsp-layout-git` from the AUR as well and (3) the install from source with `make` :thinking: 

The idea would be to use the tags in the worktree to generate the exact versions automatically:
- for `bsp-layout` the tag would be directly in the release archive used by the PKGBUILD on the AUR :ok_hand:
- for `bsp-layout-git` and `make install` the version could simply be computed with something like `git describe --tags`, e.g. right now, on the tip of the `master` branch, the version given by `bsp-layout version` would be `0.0.10-5-g181d384` cause the tip of `master` is 5 commits above `0.0.10` and with hash `181d384`. If someone happens to have installed an outdated version of `bsp-layout` from source, that would be easier to catch, i think :yum: 

is that something you would like in the project? :blush:

**Reply by phenax** [reply]
> this has to be harcoded, so that is pretty error prone

No. That is just the default version for use during development. Both AUR packages use install.sh and you can see in there that the version is the last git tag.

But yes, I agree there is some room for improvement. The reason, I didn't do this inside the Makefile was because bsp-layout-git and bsp-layout will have different versions so specifying that externally made sense.

Also, I think we need to rethink the install process and maybe sunset the install.sh file completely. This would solve the problem as both of them calling make would specify the version of the package itself so no room for error.

**Reply by amtoine** [reply]
> No. That is just the default version for use during development. Both AUR packages use install.sh and you can see in there that the version is the last git tag.
>
wow... to be honest, i did not even notice that `VERSION` being defined there :sweat: 
 
> But yes, I agree there is some room for improvement. The reason, I didn't do this inside the Makefile was because bsp-layout-git and bsp-layout will have different versions so specifying that externally made sense.
> 
not sure i get the idea... are we talking about the versions of the AUR packages or the command they ship? :thinking: 

> Also, I think we need to rethink the install process and maybe sunset the install.sh file completely. This would solve the problem as both of them calling make would specify the version of the package itself so no room for error.
>
yeah... this script is quite confusing and looks, to me, quite redundant with both the makfile and the AUR packages :thinking: 

i add that refactor to the list above :wink:

**Comment by amtoine** [comment]
Hi @phenax :wave: :yum: 

in order to prepare some of the ideas above, i thought about changing the argument parsing a bit :thinking: 
the goal is not to change the behaviour of the `bsp-layout` command at all, but simply to have a slightly easier to work with main `./src/layout.sh` file :yum: 

i've written a little demo on my personal [parsing-demo](https://github.com/a2n-s/bsp-layout/blob/parsing-demo/src/parser.sh) branch :ok_hand: 

the script should be used as if `bash ./src/parser.sh` was in fact `bsp-layout`, e.g. `bash ./src/paser.sh get 2` instead of `bsp-layout get 2`
i tried to write that script such that the flags behave exactly the same way as the original `bsp-layout` code!

the main reason for [that branch](https://github.com/a2n-s/bsp-layout/blob/parsing-demo/src/parser.sh) demo to exist is that i feel not comfortable with passing `"$@"` to the functions...

hope you'll find that useful :blush: 
do not hesitate to comment on that :wink:
if you like the idea, i think that would be the next issue i'll open once the templates are up and running :muscle:

**Reply by phenax** [reply]
What if a desktop's name can contains the `:` character? Any custom parsing logic is pretty brittle and an overkill for this purpose.

> i feel not comfortable with passing "$@" to the functions

I think it's offers a lot more flexibility if each sub-command has the ability to parse the arguments it needs and perform validations. But yes, this is a tradeoff between flexibility and having a single source of truth on the cli's api. I'm fine either way since the cli is simple enough but I don't think we should have any custom parsing methods for the internal apis because of the desktop names which can break the parser.

Another thing to consider is that, there's 2 layers to this. The core sub-commands arg parsing and the layout specific arg parsing (Eg - `bsp-layout set tall -- --master-size 0.6`). It's better to leave the layout specific arg parsing inside the layout itself and maybe we can simplifiy it with getopts instead.

**Reply by amtoine** [reply]
> What if a desktop's name can contains the `:` character? Any custom parsing logic is pretty brittle and an overkill for this purpose.
> 
that is a great remark :thinking: 
i think i might have a fix for this :+1: but sure it is useful then...

> I think it's offers a lot more flexibility if each sub-command has the ability to parse the arguments it needs and perform validations. But yes, this is a tradeoff between flexibility and having a single source of truth on the cli's api. I'm fine either way since the cli is simple enough but I don't think we should have any custom parsing methods for the internal apis because of the desktop names which can break the parser.
> 
> Another thing to consider is that, there's 2 layers to this. The core sub-commands arg parsing and the layout specific arg parsing (Eg - `bsp-layout set tall -- --master-size 0.6`). It's better to leave the layout specific arg parsing inside the layout itself and maybe we can simplifiy it with getopts instead.
>
not sure about which way to choose, the local parsing in each function or the global parsing and give exactly the arguments needed by the functions :thinking: 

maybe the local parsing is a good thing here though! :+1:

**Comment by amtoine** [comment]
i had another idea, regarding installing and compiling the code :yum: 

i like to print debug statements when developing, to understand the code and what it's doing! :ok_hand: 

however, we do not want such debug statements to end up in production code :thinking: so i need to remove all these each time i commit and write them again later when needed :confused: 

what about doing the following:
   - write debug statements with a predefined "anchor", e.g. `# {{DEBUG}}` at the end of the line, for instance
```bash
... some code
echo "value: $value"  # {{DEBUG}}
... some other code
```
   - compile the code, i.e. something like `sed "s|{{VERSION}}|${VERSION}|g" -i layout.sh.tmp` in the `Makefile` to either remove the `# {{DEBUG}}` statements, by default, or leave them with something like `make DEBUG=1 install` during development and debugging, i.e. with the example above `make DEBUG=0 install` or `make install` would produce
```bash
... some code
... some other code
```
whereas `make DEBUG=1 install` would leave `echo "value: $value"  # {{DEBUG}}`
 
 ?

i think that would be kinda cool, very practical, easy to implement and just a matter of finding the same convention for everyone :yum:

**Reply by phenax** [reply]
Good idea. But, I'm not a big fan of this sed templating trick being used in too many places tho since it is a little too messy. Can we abstract it all out into another script that given a set of key-value pairs, compiles the template for us? Handlebars but much simpler.

**Reply by amtoine** [reply]
coool :yum: 

i add that to the list above :ok_hand:

**Reply by amtoine** [reply]
not sure to understand what would a 
> script that given a set of key-value pairs, compiles the template for us
>
be :thinking: 

we'll probably talk about that in depth in an issue some day :wink:
